src/tFPDF/ProtectedPDF.php

```diff
@@ -0,0 +1,271 @@
+            $this->PutEncryption();
+            $this->Out('>>');
+            $this->Out('endobj');
+        }
+    }
+
+    /** {@inheritdoc} */
+    public function PutTrailer()
+    {
+        parent::PutTrailer();
+
+        if ($this->bol_encrypted) {
+            $this->Out('/Encrypt ' . $this->int_encryption_object_id . ' 0 R');
+            $this->Out('/ID [()()]');
+        }
+    }
+
+    /**
+     * Write encryption metadata to the document
+     */
+    public function PutEncryption()
+    {
+        $this->Out('/Filter /Standard');
+        $this->Out('/V 1');
+        $this->Out('/R 2');
+
+        $this->Out('/O ('.$this->EscapeString($this->str_o_value).')');
+        $this->Out('/U ('.$this->EscapeString($this->str_u_value).')');
+        $this->Out('/P '.$this->int_p_value);
+    }
+
+    /**
+     * @param array|null $permissions An array of ::PERMISSION_*
+     * @param string $userPassword The password for restricted access to the document (obeying $permissions)
+     * @param string|null $ownerPassword The password for unrestricted access to the document (ignoring $permissions)
+     */
+    public function SetProtection(array $permissions = null, $userPassword = '', $ownerPassword = null)
+    {
+        $protection = self::PROTECTION_BASE;
+        foreach((array)$permissions as $permission){
+            $protection += $permission;
+        }
+
+        if (strlen($userPassword) > self::PASSWORD_LENGTH || strlen((string)$ownerPassword) > self::PASSWORD_LENGTH) {
+            throw new \InvalidArgumentException('Passwords must be no greater than '.self::PASSWORD_LENGTH.' chars');
+        }
+
+        if ($ownerPassword === null) {
+            // Generate random ownerpassword
+            $ownerPassword = substr(uniqid(mt_rand(), true).uniqid(mt_rand(), true), 0, self::PASSWORD_LENGTH);
+        }
+
+        $this->InitializeEncryption($userPassword, $ownerPassword, $protection);
+
+        $this->bol_encrypted = true;
+    }
+
+    /**
+     * Compute encryption key
+     */
+    private function InitializeEncryption($userPassword, $ownerPassword, $protection)
+    {
+        // Pad passwords
+        $userPassword  = self::FormatEncryptionPassword($userPassword);
+        $ownerPassword = self::FormatEncryptionPassword($ownerPassword);
+
+        // Compute values
+        $this->str_o_value        = $this->GenerateOvalue($userPassword, $ownerPassword);
+        $this->str_encryption_key = $this->GenerateEncryptionKey($userPassword, $this->str_o_value, $protection);
+        $this->str_u_value        = $this->GenerateUvalue($this->str_encryption_key);
+        $this->int_p_value        = $this->GeneratePvalue($protection);
+    }
+
+    /**
+     * @param string $object_index
+     * @return string The key for the specified object
+     */
+    private function ObjectKey($object_index)
+    {
+        return substr($this->MD5to16($this->str_encryption_key.pack('VXxx', $object_index)), 0, 10);
+    }
```

Comment:
For the sake of consistency I think it would be better to call the parameter $objectIndex